### ISSUE TITLE
Content reporting

### ISSUE THREAD

**Issue by rafalp** [rafalp/Misago#484]
There should be a way to flag posts/private thread posts and user profiles for moderators to address.

**Comment by CyberFox001** [comment]
Hello,

At our association, we plan to use Misago as our forum.

We will be interested to develop this feature and do it as an upstream contribution.

Is this features still wanted ?

If yes, do you have any requirement ?

**Comment by rafalp** [comment]
Hey, thanks for offering help here. I was thinking about implementing reports only after I've  [simplified user permissions](https://misago-project.org/t/simplifying-user-permissions/1097/) but that's not a blocker.

My plan for this feature was to first research current state of art for reporting content. For example, some places use "report" but other use "flag". Which one should Misago use?

Next thing to research would be, what content should support reporting? Is it good enough to report posts only? But what if I spot a user with offensive avatar in search result, who never posted? Then I have to post a public thread or reach with private message to one of forum staff. Option for reporting users would be useful too.

So likely we would want to have following models:

- `ReportReason`: id, weight, content type (post or user), message, description (optional)
- `Report`: id, reporter_id (fkey to user), reporter_name, created_at, reason_id, status (new, resolved as valid, resolved as invalid), resolver_id, resolver_name, resolved_at
- `PostReport`: inherits from Report, but adds category_id, thread_id, post_id
- `UserReport`: inherits from Report, but adds `user_id` for reported user

Extra questions:

- What should happen when report reason is deleted by admin but it has reports?
- How long should reports be stored?
- What states should reports have?
- How should forum notify team members about new reports?
- Should users report same content multiple times? Should there be an anti-flood limit preventing users from opening more than some reports in given time?
- What to do when moderator closes report but reporter is unhappy with resolution?

**Comment by CyberFox001** [comment]
I'm not a expert in moderation, so my answer is only suggestions.

For the name, "report" or "flag", I don't know. But "report" seems more talking to me about what the action does. Like "Ask for moderation".

On the question "What can be reported", I think "Anything a user can create". So, report posts and users accounts is a good thing.

About the models you suggest, why do you think to make reason separate from report ?

> What should happen when report reason is deleted by admin but it has reports?

Maybe auto-delete reports with no more reasons ?

> How long should reports be stored?

I think, in the first version, as long as admins or moderators don't delete themes.
Maybe we can add an action to delete all report older than a selected date ?

> What states should reports have?

Maybe:

- Submitted
- Accepted
- Closed

> How should forum notify team members about new reports?

It's a good question. On our old forum, we are notified by emails. We also have an RSS feed for all new posts and the moderator teams look at it for spam.

> Should users report same content multiple times? Should there be an anti-flood limit preventing users from opening more than some reports in given time?

Maybe an option for administrators/moderators to "silence" a user who flood the report system ?

>  What to do when moderator closes report but reporter is unhappy with resolution?

Maybe an option for the user to ask a second opinion from another moderator ? In this case, it would require 2  closes from 2 different moderators


When a report is accepted, I think of multiple actions that can be done by administrator/moderators directly from the report:

- Delete the message
- Delete the reported account and, optionality, all it's messages
- Suspend the reported account for some time
- Silence the reporter account for some time or for ever

**Comment by rafalp** [comment]
> About the models you suggest, why do you think to make reason separate from report?

Because that way you have cheap filtering by standard report reason in potential admin panel.

> Maybe auto-delete reports with no more reasons?

IMHO reasons and reports should be one to many relation (report would have foreign key to a reason), and deleting a reason should either be prohibited if there are reports for given reason, or should delete reports together with a reason.

**Comment by CyberFox001** [comment]
> Because that way you have cheap filtering by standard report reason in potential admin panel.

Ah, `ReportReason` are the category of a report ?

> Seems ok, but should only apply for "resolved" reports. And should be an opt-in.

I was thinking of a manual action, triggered by moderators with a button in reports list UI. Do you think of an automated action, run by cron ?

>     Submitted
>     Accepted
>     Closed
> 
> What would those statuses mean?

- Submitted: When a user create the report
- Accepted: When an administrator accepted the report and take action
- Closed: When the report is refused and no action have been taken

> This is out of scope of reports IMHO. Misago already has moderator tools for user content and accounts, so its up to moderator to decide what they want to do after seeing the report.

Yes, but on the UX side, it's better if you can do the moderation actions in relation to the report from the report page or the reports list page ?


I was thinking about 2 other points.

On the user side: Users need a feedback about their reports. So, they would need an UI where they can see all their own reports, their status and what action have been taken. And if a moderator refuse a report, user need to know why.

What moderators do when the report made by user is unclear and need more explanation ? Or when moderators need explanations from reported user ? Do we add a button "Start a discussion in PM with AAA user" ?

**Comment by rafalp** [comment]
> Ah, ReportReason are the category of a report ?

Correct, basically those:

<img width="439" alt="Zrzut ekranu 2023-05-22 o 15 13 46" src="https://github.com/rafalp/Misago/assets/750553/f4ce1044-8549-4dcb-9a2a-91909c55f93b">

> Submitted: When a user create the report
> Accepted: When an administrator accepted the report and take action
> Closed: When the report is refused and no action have been taken

I would use `NEW`, `RESOLVED`, `REJECTED` for new, handled by moderator and rejected by moderator as invalid. This is how Misago did it in past.

> Yes, but on the UX side, it's better if you can do the moderation actions in relation to the report from the report page or the reports list page ?

Sadly, this is a massive amount of work as every moderation action in Misago would need to be connected to content reporting system.

> On the user side: Users need a feedback about their reports. So, they would need an UI where they can see all their own reports, their status and what action have been taken. And if a moderator refuse a report, user need to know why.
> 
> What moderators do when the report made by user is unclear and need more explanation ? Or when moderators need explanations from reported user ? Do we add a button "Start a discussion in PM with AAA user" ?

Let's not think about those for MVP. I think most forums have so little activity forum staff and posters can reach to each other via private thread or asking in public why something is or isn't done.
